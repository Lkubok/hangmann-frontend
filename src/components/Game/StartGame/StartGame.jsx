import React, { Component } from "react";
import { connect } from "react-redux";
import { compose } from "redux";
import * as Yup from "yup";
import { launchNewGame, setUserParams } from "../../../actions/gameActions";
import { withFormik, Form, Field } from "formik";

import * as selectors from "../../../reducers/selectors";
import Loading from "../../EditQuote/Loading";
import axios from "axios";

import "./StartGame.scss";
const { REACT_APP_API_HOST } = process.env;

export class StartGame extends Component {
  constructor(props) {
    super(props);
    this.state = {
      langs: []
    };
  }
  handleSubmitting = e => {
    e.preventDefault();
    const { launchNewGame, setUserParams, isValid } = this.props;
    const { email, username, lang, level } = this.props.values;

    if (isValid && username) {
      launchNewGame(username, level, lang);
      setUserParams(username, level, lang, email);
    }
  };
  componentDidMount() {
    axios
      .get(REACT_APP_API_HOST + "/quotes/langs")
      .then(response => response.data)
      .then(data => this.setState({ langs: data }));
  }
  renderSelectOptions = () => {
    return this.state.langs.map(el => (
      <option key={el} value={el}>
        {el}
      </option>
    ));
  };
  renderLoading() {
    const { errors, touched, isRequesting } = this.props;

    return isRequesting ? (
      <Loading />
    ) : (
      <div className="form-holder">
        <Form className="user-params-form" onSubmit={this.handleSubmitting}>
          <p className="game-select-label">
            User name: <span>{touched.username && errors.username}</span>
          </p>
          <Field type="text" name="username" />
          <p className="game-select-label">
            Email: <span>{touched.email && errors.email}</span>
          </p>
          <Field type="email" name="email" />
          <p className="game-select-label">
            Language: <span>{touched.lang && errors.lang}</span>
          </p>
          <Field component="select" name="lang">
            <option>--- select lang ---</option>
            {this.renderSelectOptions()}
          </Field>
          <p className="game-select-label">
            Level: <span>{touched.level && errors.level}</span>
          </p>
          <Field component="select" name="level">
            <option>--- select level ---</option>
            <option value="easy">easy</option>
            <option value="medium">medium</option>
            <option value="hard">hard</option>
          </Field>
          <button disabled={isRequesting}>Start Game</button>
        </Form>
      </div>
    );
  }
  render() {
    return <>{this.renderLoading()}</>;
  }
}

const MyForm = withFormik({
  mapPropsToValues({
    username,
    email,
    lang,
    level,
    getUserName,
    getUserEmail,
    getUserLang,
    getUserLevel
  }) {
    return {
      username: username || getUserName,
      email: email || getUserEmail,
      lang: lang || getUserLang,
      level: level || getUserLevel
    };
  },
  validationSchema: Yup.object().shape({
    username: Yup.string()
      .min(3)
      .max(50)
      .required("Username is required"),
    email: Yup.string()
      .email()
      .required("E-mail is required"),
    lang: Yup.string()
      .min(2)
      .max(50)
      .required("Select a language!"),
    level: Yup.string()
      .min(3)
      .max(25)
      .required("Select level!")
  }),
  isInitialValid: true
});

const mapStateToProps = state => {
  return {
    isRequesting: selectors.getIsRequesting(state),
    getUserName: selectors.getLastUserName(state),
    getUserEmail: selectors.getLastUserEmail(state),
    getUserLevel: selectors.getLastGameLevel(state),
    getUserLang: selectors.getLastGameLang(state)
  };
};

const mapDispatchToProps = {
  launchNewGame,
  setUserParams
};

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  MyForm
)(StartGame);
