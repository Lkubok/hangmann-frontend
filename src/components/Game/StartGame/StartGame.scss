@import "../../../scss/mixins.scss";
@import "../../../scss/variables.scss";

.loading-holder {
  @include flex-cc-row;
  @include loading-content;
  align-items: flex-start;
  border: 0;
}

.user-params-form {
  @include flex-cc-row;
  @include form-edit;
  color: $white-color;
  input,
  select {
    @include input-textarea-look;
  }
  button {
    @include std-button;
  }
}

.game-select-label {
  color: $primary-color;
  span {
    color: $danger-color;
  }
}
