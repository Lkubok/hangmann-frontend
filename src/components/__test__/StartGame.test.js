import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { StartGame } from "../Game/StartGame/StartGame";
Enzyme.configure({ adapter: new Adapter() });

describe("StartGame component", () => {
  let props;
  beforeEach(() => {
    props = {
      username: "test",
      email: "test",
      lang: "test",
      level: "test",
      getUserName: "test",
      getUserEmail: "test",
      getUserLang: "test",
      getUserLevel: "test",
      isLogged: "test",
      AppUserEmail: "test",
      AppUserName: "test",
      touched: {
        username: "test",
        email: "test",
        lang: "test",
        level: "test"
      },
      errors: {
        username: "test",
        email: "test",
        lang: "test",
        level: "test"
      }
    };
  });
  it("Should match snapshot", () => {
    const wrapper = shallow(<StartGame {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
