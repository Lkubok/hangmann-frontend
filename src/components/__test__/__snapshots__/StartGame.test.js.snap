// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartGame component Should match snapshot 1`] = `
<Fragment>
  <div
    className="form-holder"
  >
    <Form
      className="user-params-form"
      onSubmit={[Function]}
    >
      <p
        className="game-select-label"
      >
        User name: 
        <span>
          test
        </span>
      </p>
      <FormikConnect(FieldInner)
        name="username"
        type="text"
      />
      <p
        className="game-select-label"
      >
        Email: 
        <span>
          test
        </span>
      </p>
      <FormikConnect(FieldInner)
        name="email"
        type="email"
      />
      <p
        className="game-select-label"
      >
        Language: 
        <span>
          test
        </span>
      </p>
      <FormikConnect(FieldInner)
        component="select"
        name="lang"
      >
        <option>
          --- select lang ---
        </option>
      </FormikConnect(FieldInner)>
      <p
        className="game-select-label"
      >
        Level: 
        <span>
          test
        </span>
      </p>
      <FormikConnect(FieldInner)
        component="select"
        name="level"
      >
        <option>
          --- select level ---
        </option>
        <option
          value="easy"
        >
          easy
        </option>
        <option
          value="medium"
        >
          medium
        </option>
        <option
          value="hard"
        >
          hard
        </option>
      </FormikConnect(FieldInner)>
      <button>
        Start Game
      </button>
    </Form>
  </div>
</Fragment>
`;
